File: Services/ImageCacheService.swift
import UIKit
import Foundation

class ImageCacheService {
    static let shared = ImageCacheService()
    
    private let memoryCache = NSCache<NSString, UIImage>()
    private let fileManager = FileManager.default
    private let cacheDirectory: URL
    
    private init() {
        // Utiliser le dossier App Group pour partager avec les widgets
        if let containerURL = fileManager.containerURL(forSecurityApplicationGroupIdentifier: "group.com.lyes.love2love") {
            cacheDirectory = containerURL.appendingPathComponent("ImageCache")
        } else {
            // Fallback vers le cache Documents de l'app
            cacheDirectory = fileManager.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("ImageCache")
        }
        
        // Créer le dossier de cache s'il n'existe pas
        try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        
        // Configuration du cache mémoire
        memoryCache.countLimit = 50 // Maximum 50 images en mémoire
        memoryCache.totalCostLimit = 100 * 1024 * 1024 // Maximum 100MB en mémoire
        
        print("🖼️ ImageCacheService: Cache initialisé - Dossier: \(cacheDirectory.path)")
    }
    
    // MARK: - Public Methods
    
    func getCachedImage(for urlString: String) -> UIImage? {
        let cacheKey = cacheKeyForURL(urlString)
        
        // 1. Vérifier le cache mémoire d'abord (plus rapide)
        if let memoryImage = memoryCache.object(forKey: cacheKey as NSString) {
            // Log sécurisé sans exposer l'URL complète avec token
            print("🖼️ ImageCacheService: Image trouvée en cache mémoire")
            return memoryImage
        }
        
        // 2. Vérifier le cache disque
        if let diskImage = loadImageFromDisk(cacheKey: cacheKey) {
            // Log sécurisé sans exposer l'URL complète avec token
            print("🖼️ ImageCacheService: Image trouvée en cache disque")
            // Remettre en cache mémoire
            memoryCache.setObject(diskImage, forKey: cacheKey as NSString)
            return diskImage
        }
        
        // Log sécurisé sans exposer l'URL complète avec token
        print("🖼️ ImageCacheService: Aucune image en cache")
        return nil
    }
    
    func cacheImage(_ image: UIImage, for urlString: String) {
        let cacheKey = cacheKeyForURL(urlString)
        
        // 1. Mettre en cache mémoire
        memoryCache.setObject(image, forKey: cacheKey as NSString)
        
        // 2. Mettre en cache disque de façon asynchrone
        Task.detached { [weak self] in
            self?.saveImageToDisk(image, cacheKey: cacheKey)
        }
        
        // Log sécurisé sans exposer l'URL complète avec token
        print("🖼️ ImageCacheService: Image mise en cache")
    }
    
    func clearCachedImage(for urlString: String) {
        let cacheKey = cacheKeyForURL(urlString)
        
        // 1. Supprimer du cache mémoire
        memoryCache.removeObject(forKey: cacheKey as NSString)
        
        // 2. Supprimer du cache disque
        Task.detached { [weak self] in
            guard let self = self else { return }
            let fileURL = self.cacheDirectory.appendingPathComponent("\(cacheKey).jpg")
            try? self.fileManager.removeItem(at: fileURL)
        }
        
        // Log sécurisé sans exposer l'URL complète avec token
        print("🗑️ ImageCacheService: Image supprimée du cache")
    }
    
    func clearCache() {
        // Vider le cache mémoire
        memoryCache.removeAllObjects()
        
        // Vider le cache disque
        Task.detached { [weak self] in
            guard let self = self else { return }
            try? self.fileManager.removeItem(at: self.cacheDirectory)
            try? self.fileManager.createDirectory(at: self.cacheDirectory, withIntermediateDirectories: true)
        }
        
        print("🖼️ ImageCacheService: Cache vidé")
    }
    
    func getCacheSize() -> (memorySize: String, diskSize: String) {
        let memorySize = "~\(memoryCache.totalCostLimit / 1024 / 1024)MB"
        
        var diskSize = "0MB"
        if let enumerator = fileManager.enumerator(at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey]) {
            var totalSize: Int64 = 0
            for case let fileURL as URL in enumerator {
                if let resourceValues = try? fileURL.resourceValues(forKeys: [.fileSizeKey]),
                   let fileSize = resourceValues.fileSize {
                    totalSize += Int64(fileSize)
                }
            }
            diskSize = ByteCountFormatter.string(fromByteCount: totalSize, countStyle: .file)
        }
        
        return (memorySize, diskSize)
    }
    
    // MARK: - Private Methods
    
    private func cacheKeyForURL(_ urlString: String) -> String {
        // Créer une clé unique basée sur l'URL
        let url = URL(string: urlString)
        if let path = url?.path, let query = url?.query {
            return "\(path.replacingOccurrences(of: "/", with: "_"))_\(query.hash)".replacingOccurrences(of: "=", with: "_")
        } else if let path = url?.path {
            return path.replacingOccurrences(of: "/", with: "_")
        } else {
            return urlString.hash.description
        }
    }
    
    private func loadImageFromDisk(cacheKey: String) -> UIImage? {
        let fileURL = cacheDirectory.appendingPathComponent("\(cacheKey).jpg")
        
        guard fileManager.fileExists(atPath: fileURL.path),
              let imageData = try? Data(contentsOf: fileURL),
              let image = UIImage(data: imageData) else {
            return nil
        }
        
        return image
    }
    
    private func saveImageToDisk(_ image: UIImage, cacheKey: String) {
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            print("❌ ImageCacheService: Impossible de convertir l'image en données")
            return
        }
        
        let fileURL = cacheDirectory.appendingPathComponent("\(cacheKey).jpg")
        
        do {
            try imageData.write(to: fileURL)
            // Log sécurisé sans exposer le nom de fichier avec UID
            print("🖼️ ImageCacheService: Image sauvée sur disque")
        } catch {
            print("❌ ImageCacheService: Erreur sauvegarde disque: \(error)")
        }
    }
    
    // MARK: - Widget Compatibility
    
    /// Méthode compatible avec WidgetService pour unifier les caches
    func cacheImageForWidget(_ image: UIImage, fileName: String) {
        print("🖼️ ImageCacheService: cacheImageForWidget appelé")
        print("  - FileName: \(fileName)")
        print("  - CacheDirectory: \(cacheDirectory.path)")
        
        let fileURL = cacheDirectory.appendingPathComponent(fileName)
        print("  - FileURL: \(fileURL.path)")
        
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            print("❌ ImageCacheService: Impossible de convertir l'image en JPEG")
            return
        }
        
        print("  - ImageData size: \(imageData.count) bytes")
        
        do {
            try imageData.write(to: fileURL)
            print("✅ ImageCacheService: Image widget sauvée: \(fileName)")
            
            // Vérifier que le fichier existe bien
            if FileManager.default.fileExists(atPath: fileURL.path) {
                print("✅ ImageCacheService: Vérification - Fichier existe: \(fileURL.path)")
                
                // Vérifier la taille du fichier
                if let attributes = try? FileManager.default.attributesOfItem(atPath: fileURL.path),
                   let fileSize = attributes[.size] as? NSNumber {
                    print("✅ ImageCacheService: Taille fichier: \(fileSize.intValue) bytes")
                }
            } else {
                print("❌ ImageCacheService: Vérification - Fichier n'existe pas après sauvegarde!")
            }
        } catch {
            print("❌ ImageCacheService: Erreur sauvegarde widget: \(error)")
            print("❌ ImageCacheService: Error details: \(error.localizedDescription)")
        }
    }
    
    func getCachedWidgetImage(fileName: String) -> UIImage? {
        let fileURL = cacheDirectory.appendingPathComponent(fileName)
        
        guard let imageData = try? Data(contentsOf: fileURL),
              let image = UIImage(data: imageData) else {
            return nil
        }
        
        return image
    }
} 

File: Services/ImageCacheService.kt
package com.lyes.love2love.cache

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Log
import android.util.LruCache
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.text.DecimalFormat
import kotlin.math.max

/**
 * ImageCacheService (Android)
 * - Cache mémoire via LruCache (limité en octets)
 * - Cache disque dans <app>/cache/ImageCache
 * - Methods : getCachedImage, cacheImage, clearCachedImage, clearCache, getCacheSize
 * - Compatibilité widgets : cacheImageForWidget, getCachedWidgetImage
 *
 * Usage :
 *   val cache = ImageCacheService.getInstance(context)
 *   cache.cacheImage(bitmap, url)
 *   val bmp = cache.getCachedImage(url)
 */
class ImageCacheService private constructor(
    private val appContext: Context
) {

    companion object {
        private const val TAG = "ImageCacheService"

        // Limites mémoire (alignées sur ton Swift : ~100 MB)
        private const val MEMORY_MAX_BYTES: Int = 100 * 1024 * 1024 // 100 MB
        // Swift limitait à 50 images, LruCache ne gère pas un "countLimit" nativement.
        // On se repose ici sur la limite en octets.

        @Volatile
        private var INSTANCE: ImageCacheService? = null

        fun getInstance(context: Context): ImageCacheService =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: ImageCacheService(context.applicationContext).also { INSTANCE = it }
            }
    }

    // --- Mémoire ---
    private val memoryCache = object : LruCache<String, Bitmap>(MEMORY_MAX_BYTES) {
        override fun sizeOf(key: String, value: Bitmap): Int {
            // Compte la taille réelle du bitmap
            return value.byteCount
        }
    }

    // --- Disque ---
    private val cacheDirectory: File = File(appContext.cacheDir, "ImageCache").apply {
        if (!exists()) mkdirs()
    }

    // --- Concurrence disque ---
    private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    init {
        Log.i(TAG, "🖼️ Cache initialisé — Dossier: ${cacheDirectory.absolutePath}")
    }

    // =====================
    // Public API
    // =====================

    /** Retourne un Bitmap si présent en cache mémoire ou disque, sinon null. */
    fun getCachedImage(urlString: String): Bitmap? {
        val key = cacheKeyForURL(urlString)

        // 1) Mémoire
        memoryCache.get(key)?.let {
            Log.d(TAG, "🖼️ Image trouvée en cache mémoire")
            return it
        }

        // 2) Disque
        loadImageFromDisk(key)?.let { diskBmp ->
            Log.d(TAG, "🖼️ Image trouvée en cache disque")
            memoryCache.put(key, diskBmp) // Remonte en mémoire
            return diskBmp
        }

        Log.d(TAG, "🖼️ Aucune image en cache")
        return null
    }

    /** Ajoute/Met à jour une image en cache (mémoire + disque asynchrone). */
    fun cacheImage(image: Bitmap, urlString: String) {
        val key = cacheKeyForURL(urlString)

        // Mémoire
        memoryCache.put(key, image)

        // Disque (asynchrone)
        ioScope.launch {
            saveImageToDisk(image, key)
        }

        Log.d(TAG, "🖼️ Image mise en cache")
    }

    /** Supprime une image spécifique du cache (mémoire + disque). */
    fun clearCachedImage(urlString: String) {
        val key = cacheKeyForURL(urlString)

        // Mémoire
        memoryCache.remove(key)

        // Disque
        ioScope.launch {
            val file = File(cacheDirectory, "$key.jpg")
            if (file.exists()) file.delete()
        }

        Log.d(TAG, "🗑️ Image supprimée du cache")
    }

    /** Vide tout le cache (mémoire + disque). */
    fun clearCache() {
        memoryCache.evictAll()
        ioScope.launch {
            cacheDirectory.deleteRecursively()
            cacheDirectory.mkdirs()
        }
        Log.i(TAG, "🖼️ Cache vidé")
    }

    /** Taille du cache (mémoire théorique max + disque réel). */
    fun getCacheSize(): CacheSize {
        val memorySizeHuman = "~${MEMORY_MAX_BYTES / 1024 / 1024}MB"

        var diskBytes = 0L
        cacheDirectory.walkTopDown().forEach { f ->
            if (f.isFile) diskBytes += f.length()
        }
        val diskSizeHuman = humanReadableByteCount(diskBytes)

        return CacheSize(memorySizeHuman, diskSizeHuman)
    }

    // =====================
    // Widget Compatibility
    // =====================

    /**
     * Sauvegarde synchrone pour widgets (nom de fichier fourni).
     * Remarque : sur Android, le widget ne lit pas directement le fichier ;
     * c’est l’app qui charge l’image et l’assigne au RemoteViews/Glance.
     */
    fun cacheImageForWidget(image: Bitmap, fileName: String) {
        Log.d(TAG, "🖼️ cacheImageForWidget(file=$fileName)")
        val file = File(cacheDirectory, fileName)
        try {
            FileOutputStream(file).use { out ->
                image.compress(Bitmap.CompressFormat.JPEG, 80, out)
            }
            if (file.exists()) {
                Log.d(TAG, "✅ Image widget sauvée: ${file.name} (${humanReadableByteCount(file.length())})")
            } else {
                Log.e(TAG, "❌ Fichier widget introuvable après écriture")
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Erreur sauvegarde widget: ${e.message}", e)
        }
    }

    /** Charge un bitmap destiné à un widget à partir d’un nom de fichier. */
    fun getCachedWidgetImage(fileName: String): Bitmap? {
        val file = File(cacheDirectory, fileName)
        if (!file.exists()) return null
        return try {
            BitmapFactory.decodeStream(FileInputStream(file))
        } catch (e: Exception) {
            Log.e(TAG, "❌ Erreur lecture image widget: ${e.message}", e)
            null
        }
    }

    // =====================
    // Privates
    // =====================

    private fun cacheKeyForURL(urlString: String): String {
        // Clé stable et “safe” pour système de fichiers
        return try {
            val uri = Uri.parse(urlString)
            val path = (uri.path ?: urlString).replace("/", "_")
            val queryHash = uri.query?.hashCode()?.toString()
            val base = if (!queryHash.isNullOrEmpty()) "${path}_$queryHash" else path

            // Nettoyage des caractères problématiques + limite longueur
            val cleaned = base.replace(Regex("[^A-Za-z0-9._-]"), "_")
            if (cleaned.length <= 200) cleaned
            else cleaned.take(150) + "_" + cleaned.hashCode()
        } catch (_: Exception) {
            // Fallback ultra-robuste
            urlString.hashCode().toString()
        }
    }

    private fun loadImageFromDisk(cacheKey: String): Bitmap? {
        val file = File(cacheDirectory, "$cacheKey.jpg")
        if (!file.exists()) return null
        return try {
            BitmapFactory.decodeStream(FileInputStream(file))
        } catch (e: Exception) {
            Log.e(TAG, "❌ Erreur lecture disque: ${e.message}", e)
            null
        }
    }

    private fun saveImageToDisk(image: Bitmap, cacheKey: String) {
        val file = File(cacheDirectory, "$cacheKey.jpg")
        try {
            FileOutputStream(file).use { out ->
                image.compress(Bitmap.CompressFormat.JPEG, 80, out)
            }
            Log.d(TAG, "🖼️ Image sauvée sur disque")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Erreur sauvegarde disque: ${e.message}", e)
        }
    }

    private fun humanReadableByteCount(bytes: Long): String {
        if (bytes < 1024) return "$bytes B"
        val unit = 1024.0
        val exp = (Math.log(bytes.toDouble()) / Math.log(unit)).toInt()
        val pre = "KMGTPE"[exp - 1]
        val df = DecimalFormat("#.#")
        return "${df.format(bytes / Math.pow(unit, exp.toDouble()))} ${pre}B"
    }

    data class CacheSize(
        val memorySize: String,
        val diskSize: String
    )
}
